### ISSUE TITLE
LayerNorm+CUDA+JIT

### ISSUE THREAD

**Issue by DavidSlayback** [pytorch/pytorch#82889]
### 🐛 Describe the bug

[File](https://gist.github.com/DavidSlayback/dd327d17d41e67267f0026efb0686fc5)

In an RL workflow, when I use CUDA, JIT, and a LayerNorm together in this script, I get the following error
```python

/home/david/Documents/GitHub/cleanoc/cleanrl/ppo_lngru_jit.py:281: UserWarning: FALLBACK path has been taken inside: compileCudaFusionGroup. This is an indication that codegen Failed for some reason.
To debug try disable codegen fallback path via setting the env variable `export PYTORCH_NVFUSER_DISABLE=fallback`
To report the issue, try enable logging via setting the envvariable ` export PYTORCH_JIT_LOG_LEVEL=manager.cpp`
 (Triggered internally at  /opt/conda/conda-bld/pytorch_1656352657443/work/torch/csrc/jit/codegen/cuda/manager.cpp:237.)
  next_gru_state = agent.get_next_state(next_obs, next_gru_state, next_done, prev_actions[step])
/home/david/Documents/GitHub/cleanoc/cleanrl/ppo_lngru_jit.py:281: UserWarning: FALLBACK path has been taken inside: runCudaFusionGroup. This is an indication that codegen Failed for some reason.
To debug try disable codegen fallback path via setting the env variable `export PYTORCH_NVFUSER_DISABLE=fallback`
 (Triggered internally at  /opt/conda/conda-bld/pytorch_1656352657443/work/torch/csrc/jit/codegen/cuda/manager.cpp:329.)
  next_gru_state = agent.get_next_state(next_obs, next_gru_state, next_done, prev_actions[step])
Traceback (most recent call last):
  File "/home/david/Documents/GitHub/cleanoc/cleanrl/ppo_lngru_jit.py", line 281, in <module>
    next_gru_state = agent.get_next_state(next_obs, next_gru_state, next_done, prev_actions[step])
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: Expected weight to be of same shape as normalized_shape, but got weight of shape [4, 128] and normalized_shape = [384]
```

Following the advice above (manager.cpp for debugging, disable NVFUSER), I get
```python
Traceback (most recent call last):
  File "/home/david/Documents/GitHub/cleanoc/cleanrl/ppo_lngru_jit.py", line 277, in <module>
    next_gru_state = agent.get_next_state(next_obs, next_gru_state, next_done, prev_actions[step])
RuntimeError: Invalid broadcast, number of false entries in is_broadcast_dim expected to be 2 but received 1
```

This does not occur if:

1. I run on CPU
2. I uncomment any of the print statements inside the LNGRU
3. I turn off the first layer norms
4. I assert that both i_n and i_h have a shape (any shape that tests True)
5. I attempt to reproduce with JUST the LNGRU or even the Agent (I tried making a minimal working example with both but could not reproduce the bug). See below:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class LNGRUCell(nn.RNNCellBase):
    n_preact: torch.jit.Final[bool]
    """Layer-normalized GRU as in https://arxiv.org/pdf/1607.06450.pdf
    https://github.com/pytorch/pytorch/issues/12482#issuecomment-440485163"""

    def __init__(self, input_size, hidden_size, bias=True, n_preact=True):
        super().__init__(input_size, hidden_size, bias, num_chunks=3)
        self.n_preact = n_preact
        if n_preact:
            self.n_ih = nn.LayerNorm(int(3 * self.hidden_size))
            self.n_hh = nn.LayerNorm(int(3 * self.hidden_size))
        self.n_in = nn.LayerNorm(self.hidden_size)
        self.n_hn = nn.LayerNorm(self.hidden_size)
        # Orthogonal initialization
        nn.init.orthogonal_(self.weight_hh, 2 ** 0.5)
        nn.init.orthogonal_(self.weight_ih, 2 ** 0.5)
        if self.bias:
            nn.init.constant_(self.bias_hh, 0)
            nn.init.constant_(self.bias_ih, 0)

    def forward(self, x, gru_state):
        ih = x @ self.weight_ih.T + self.bias_ih
        hh = gru_state @ self.weight_hh.T + self.bias_hh
        if self.n_preact:  # In CUDA, with jit, breaks here
            ih = self.n_ih(ih)
            hh = self.n_hh(hh)

        i_r, i_z, i_n = ih.chunk(3, dim=1)
        h_r, h_z, h_n = hh.chunk(3, dim=1)
        # No idea why I need to do this, but ok...
        # assert i_n.shape# [-1] == self.hidden_size
        # assert h_n.shape# [-1] == self.hidden_size
        i_n = self.n_in(i_n)
        h_n = self.n_hn(h_n)

        r = torch.sigmoid(i_r + h_r)
        z = torch.sigmoid(i_z + h_z)
        n = torch.tanh(i_n + r * h_n)
        h = (1 - z) * n + z * gru_state
        return h

class LNGRUCell_WithAssert(nn.RNNCellBase):
    n_preact: torch.jit.Final[bool]
    """Layer-normalized GRU as in https://arxiv.org/pdf/1607.06450.pdf
    https://github.com/pytorch/pytorch/issues/12482#issuecomment-440485163"""

    def __init__(self, input_size, hidden_size, bias=True, n_preact=True):
        super().__init__(input_size, hidden_size, bias, num_chunks=3)
        self.n_preact = n_preact
        if n_preact:
            self.n_ih = nn.LayerNorm(int(3 * self.hidden_size))
            self.n_hh = nn.LayerNorm(int(3 * self.hidden_size))
        self.n_in = nn.LayerNorm(self.hidden_size)
        self.n_hn = nn.LayerNorm(self.hidden_size)
        # Orthogonal initialization
        nn.init.orthogonal_(self.weight_hh, 2 ** 0.5)
        nn.init.orthogonal_(self.weight_ih, 2 ** 0.5)
        if self.bias:
            nn.init.constant_(self.bias_hh, 0)
            nn.init.constant_(self.bias_ih, 0)

    def forward(self, x, gru_state):
        ih = x @ self.weight_ih.T + self.bias_ih
        hh = gru_state @ self.weight_hh.T + self.bias_hh
        if self.n_preact:
            ih = self.n_ih(ih)
            hh = self.n_hh(hh)

        i_r, i_z, i_n = ih.chunk(3, dim=1)
        h_r, h_z, h_n = hh.chunk(3, dim=1)
        # No idea why I need to do this, but ok...
        assert i_n.shape
        assert h_n.shape
        i_n = self.n_in(i_n)
        h_n = self.n_hn(h_n)

        r = torch.sigmoid(i_r + h_r)
        z = torch.sigmoid(i_z + h_z)
        n = torch.tanh(i_n + r * h_n)
        h = (1 - z) * n + z * gru_state
        return h




if __name__ == "__main__":
    na, batch_size, input_size, hidden_size = 2, 8, 128, 256
    class Agent(nn.Module):
        num_actions: torch.jit.Final[int]
        def __init__(self, with_assert=False, with_preact=True):
            super().__init__()
            self.num_actions = 2
            if with_assert: self.rnn = LNGRUCell_WithAssert(input_size + na, hidden_size, n_preact=with_preact)
            self.rnn = LNGRUCell(input_size + na, hidden_size, n_preact=with_preact)

        @torch.jit.export
        def get_next_state(self, x, gru_state, is_init, prev_action):
            return self.rnn(torch.cat([x, F.one_hot(prev_action, self.num_actions)], -1),
                            (1. - is_init.unsqueeze(-1)) * gru_state)

    # Input and hidden state
    x = torch.ones((batch_size, input_size)); xc = x.to('cuda')
    a = torch.randint(2, (batch_size,), dtype=torch.int64); ac = a.to('cuda')
    is_init = torch.zeros(batch_size); is_initc = is_init.to('cuda')
    h = torch.zeros((batch_size, hidden_size)); hc = h.to('cuda')
    # Without JIT, cuda and cpu. Works!
    cuda_lngru = Agent().to('cuda')
    cpu_lngru = Agent()
    _ = cuda_lngru.get_next_state(xc, hc, is_initc, ac)
    _ = cpu_lngru.get_next_state(x, h, is_init, a)
    # With JIT, cuda and cpu. CPU works! CUDA doesn't...
    cuda_lngru = torch.jit.script(Agent().to('cuda'))
    cpu_lngru = torch.jit.script(Agent())
    _ = cpu_lngru.get_next_state(x, h, is_init, a)
    try:
        _ = cuda_lngru.get_next_state(xc, hc, is_initc, ac)
        _ = cuda_lngru.get_next_state(xc, _, is_initc, ac)
    except Exception as e:
        print(f'CUDA size error: \n{e}')
    # With JIT and assert, CUDA works
    cuda_lngru = torch.jit.script(Agent(with_assert=True).to('cuda'))
    _ = cuda_lngru.get_next_state(xc, hc, is_initc, ac)
    print('Assert works!')
    # With JIT and without the "preactivation" LayerNorm, CUDA works
    cuda_lngru = torch.jit.script(Agent(with_preact=False).to('cuda'))
    _ = cuda_lngru.get_next_state(xc, hc, is_initc, ac)
    print('No preactivate works')
```

### Versions

PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Linux Mint 20.3 (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:56:21)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-122-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.7.64
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080
Nvidia driver version: 515.43.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==1.12.0
[pip3] torchaudio==0.12.0
[pip3] torchvision==0.13.0
[conda] blas                      2.115                       mkl    conda-forge
[conda] blas-devel                3.9.0            15_linux64_mkl    conda-forge
[conda] cudatoolkit               11.6.0              hecad31d_10    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            15_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            15_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            15_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            15_linux64_mkl    conda-forge
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] pytorch                   1.12.0          py3.9_cuda11.6_cudnn8.3.2_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.12.0               py39_cu116    pytorch
[conda] torchvision               0.13.0               py39_cu116    pytorch

**Comment by davidberard98** [comment]
@DavidSlayback does the same error occur with LayerNorm+CUDA but without JIT?

**Comment by DavidSlayback** [comment]
No, only the combination of all 3, plus whatever's going on in the gist I linked. Sorry I couldn't refine it down to a smaller reproduction script :(. It occurs on the **2nd** step through the network

**Comment by ngimel** [comment]
cc @jjsjann123

**Comment by jjsjann123** [comment]
It's probably both TS and nvfuser here.

Looks like even fallback path causes failures in layer here. I'll take a better look

**Comment by jjsjann123** [comment]
Am I supposed to see the exception thrown within the try block? @DavidSlayback 

```
root@a0befa8bea8a:/jiej/playground# python repro_82899.py
Assert works!
No preactivate works
```

Doesn't seem to give me any error. I tried this on our nightlies as well as on 1.12 pytorch container. `pytorch/pytorch:1.12.0-cuda11.3-cudnn8-runtime`.

**Comment by jjsjann123** [comment]
Got it. Have the repro locally. Something went wrong in our fusion pass.. I'm working on it.

**Comment by jjsjann123** [comment]
I think there's something funny with ConstantChunk here.... We are definitely abusing that and should not have push it for non-pw operations....

**Comment by davidberard98** [comment]
@jjsjann123 so does the issue go away when we turn off nvfuser?

**Comment by jjsjann123** [comment]
> @jjsjann123 so does the issue go away when we turn off nvfuser?

It does if I run with `PYTORCH_JIT_USE_NNC_NOT_NVFUSER=1 python repro_82899.py --cuda True`

Does NNC default not fusing normalizations? Is NNC using the same ConstantChunk optimization?

**Comment by davidberard98** [comment]
ConstantChunk is inserted as part of CanonicalizeOps, which I think will happen either way: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp#L340

Is that the optimization you're talking about or is there another detail ConstantChunk that you're referring to?

**Comment by jjsjann123** [comment]
I was referring to this https://github.com/pytorch/pytorch/blob/aad3b8e4d3b00f9bd95bc52fd5242dd0f1c43557/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L485-L503

nvfuser doesn't fuse ConstantChunk, instead we rely on this old optimization pass that we stole from legacy fuser: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/graph_fuser.cpp#L875

I patched it in #83083

**Comment by jjsjann123** [comment]
FYI, a simpler repro:

```
import torch
import torch.nn as nn
import torch.nn.functional as F

class Cell(nn.Module):
    """Layer-normalized GRU as in https://arxiv.org/pdf/1607.06450.pdf
    https://github.com/pytorch/pytorch/issues/12482#issuecomment-440485163"""

    def __init__(self):
        super().__init__()
        self.layer0 = nn.LayerNorm(384)
        self.layer1 = nn.LayerNorm(128)

    def forward(self, x, y):
        ih = self.layer0(x)
        i_r, i_z, i_n = ih.chunk(3, dim=1)
        i_n = self.layer1(i_n)
        r = torch.sigmoid(i_r)
        n = torch.tanh(i_n + r * i_z)
        h = n + r * y
        return h

if __name__ == "__main__":

    with torch.no_grad():
        # Input and hidden state
        x = torch.ones((4, 384)).to('cuda')
        h = torch.zeros((4, 128)).to('cuda')

        cuda_cell = torch.jit.script(Cell().to('cuda'))
        for _ in range(5):
            cuda_cell(x, h)
```

I'll put this in a python test.
